File: app/src/main/java/com/lence/penguinsvskillerwhales/model/Orca.java
package com.lence.penguinsvskillerwhales.model;

import android.graphics.Point;

import com.lence.penguinsvskillerwhales.view.UpdatePresenter;

public class Orca implements Organism {
    private final int lifeTime = 8;
    private final int limitHunger = 3;

    private Boolean moved = false;
    private int hunger = 0;
    private int age = 0;
    private UpdatePresenter updatePresenter;

    public Orca(UpdatePresenter updatePresenter) {
        this.updatePresenter = updatePresenter;
    }

    private int getHunger() {
        return hunger;
    }

    public void addHunger() {
        hunger++;
    }

    public void eat() {
        hunger = 0;
    }

    private void addAge() {
        if (age < lifeTime) {
            age++;
        } else {
            age = 1;
        }
    }

    public void lifeCycle(Point point) {
        if (!isMoved()) {
            if (getHunger() < limitHunger-1) {
                if (getAge() == lifeTime-1) {
                    updatePresenter.breeding(point, new Orca(updatePresenter));
                }
                updatePresenter.eat(point, this);
                addAge();
                setMoved(true);
            } else {
                updatePresenter.death(point);

            }
        }
    }

    public int getAge() {
        return age;
    }

    @Override
    public Boolean isMoved() {
        return moved;
    }

    @Override
    public void setMoved(Boolean moved) {
        this.moved = moved;
    }
}

File: app/src/main/java/com/lence/penguinsvskillerwhales/model/Orca.kt
package com.lence.penguinsvskillerwhales.model

import android.graphics.Point

import com.lence.penguinsvskillerwhales.view.UpdatePresenter

class Orca(private val updatePresenter: UpdatePresenter) : Organism {
    private val lifeTime = 8
    private val limitHunger = 3

    override var moved: Boolean = false
    private var hunger = 0
    override var age = 0

    fun addHunger() {
        hunger++
    }

    fun eat() {
        hunger = 0
    }

    private fun addAge() {
        if (age < lifeTime) {
            age++
        } else {
            age = 1
        }
    }

    override fun lifeCycle(point: Point) {
        if ((!moved)) {
            if (hunger < limitHunger - 1) {
                if (age == lifeTime - 1) {
                    updatePresenter.breeding(point, Orca(updatePresenter))
                }
                updatePresenter.eat(point, this)
                addAge()
                moved = true
            } else {
                updatePresenter.death(point)

            }
        }
    }


}
